src/Data/Aeson/TextMap.hs

```diff
@@ -0,0 +1,230 @@
+
+
+instance TH.Lift v => TH.Lift (TextMap v) where
+    lift (TextMap m) = [| TextMap (H.fromList . map (first pack) $ m') |]
+        where
+          m' = map (first unpack) . H.toList $ m
+
+#if MIN_VERSION_template_haskell(2,17,0)
+    liftTyped = TH.unsafeCodeCoerce . TH.lift
+#elif MIN_VERSION_template_haskell(2,16,0)
+    liftTyped = TH.unsafeTExpCoerce . TH.lift
+#endif
+
+-- |
+-- Construct an empty map.
+empty :: TextMap v
+empty = TextMap H.empty
+
+-- |
+-- Return the number of key-value mappings in this map.
+size :: TextMap v -> Int
+size = H.size . unTextMap
+
+-- |
+-- Construct a map with a single element.
+singleton :: Text -> v -> TextMap v
+singleton k v = TextMap (H.singleton k v)
+
+member :: Text -> TextMap a -> Bool
+member t (TextMap m) = H.member t m
+
+-- | Return the value to which the specified key is mapped,
+-- or Nothing if this map contains no mapping for the key.
+lookup :: Text -> TextMap v -> Maybe v
+lookup t tm = H.lookup t (unTextMap tm)
+
+-- | Associate the specified value with the specified key
+-- in this map. If this map previously contained a mapping
+-- for the key, the old value is replaced.
+insert :: Text -> v -> TextMap v -> TextMap v
+insert k v tm = TextMap (H.insert k v (unTextMap tm))
+
+-- | Reduce this map by applying a binary operator to all
+-- elements, using the given starting value (typically the
+-- right-identity of the operator).
+foldrWithKey :: (Text -> v -> a -> a) -> a -> TextMap v -> a
+foldrWithKey f a = H.foldrWithKey f a . unTextMap
+
+-- | Perform an Applicative action for each key-value pair
+-- in a 'TextMap' and produce a 'TextMap' of all the results.
+traverseWithKey :: Applicative f => (Text -> v1 -> f v2) -> TextMap v1 -> f (TextMap v2)
+traverseWithKey f = fmap TextMap . H.traverseWithKey f  . unTextMap
+
+-- | Construct a map from a list of elements. Uses the
+-- provided function, f, to merge duplicate entries with
+-- (f newVal oldVal).
+fromListWith :: (v -> v -> v) ->  [(Text, v)] -> TextMap v
+fromListWith op = TextMap . H.fromListWith op
+
+-- |  Construct a map with the supplied mappings. If the
+-- list contains duplicate mappings, the later mappings take
+-- precedence.
+fromList :: [(Text, v)] -> TextMap v
+fromList = TextMap . H.fromList
+
+-- | Return a list of this map's elements.
+toList :: TextMap v -> [(Text, v)]
+toList = H.toList . unTextMap
+
+-- | Return a list of this map's elements in ascending order
+-- based of the textual key.
+toAscList :: TextMap v -> [(Text, v)]
+toAscList = sortBy (comparing fst) . toList
+
+-- | Difference of two maps. Return elements of the first
+-- map not existing in the second.
+difference :: TextMap v -> TextMap v' -> TextMap v
+difference tm1 tm2 = TextMap (H.difference (unTextMap tm1) (unTextMap tm2))
+
+-- | Return a list of this map's keys.
+keys :: TextMap v -> [Text]
+keys = H.keys . unTextMap
+
+-- | Convert a 'TextMap' to a 'HashMap'.
+toHashMap :: TextMap v -> HashMap Text v
+toHashMap = unTextMap
```

Comment:
Why no inline here?

Reply:
I didn't add inline anywhere, I think, in general, it is best practice to add inline when a benchmark actually proves it is worthwhile and not before, but I am open to changing it if others agree.

Reply:
I'm thinking how get rid (or at least discourage the usage) of this function. We don't want to have `HashMap` intermediates, if they are not necessary i.e. we are literally parsing into `HashMap` / serializing from `HashMap`...

Reply:
Let me investigate if these functions are really needed. For the instances for HashMap itself we can probably use a fold which will be more representative of an abstract interface.

Reply:
They are fine for now.

Reply:
To clarify: we would like to avoid rebuilding a *Map when the representation is the same. nice to have.

Reply:
There might be some way to not expose it as part of the interface but add a re-write rule for this representation or something like that.